### PR TITLE
SPIR-V backend scaffolding

### PR DESCRIPTION
**Pull request by Jasper-Bekkers**
As discussed on Zulip, we're starting to ramp up our efforts on providing the SPIR-V backend, so I thought it might be a good time to provide some insights into the project to the wasmtime team. 

I figure it's easier for everybody if we engage early on rather then submit a giant PR in the end, this way we can take it together and one step at a time.

We'll most likely have some changes incoming for internal cranelift datastructures etc to be able to deal with SPIR-V properly but they're not part of this PR yet, we'll try and contribute those as smaller improvements.

Right now the PR is a draft because we'd like to get initial feedback on our approach, because it still contains some debug code and because I need to rebase this on top of master.

### CODE REVIEW

File: cranelift/codegen/Cargo.toml [hunk 1]
```diff
@@ -26,6 +26,9 @@ thiserror = "1.0.4"
 byteorder = { version = "1.3.2", default-features = false }
 peepmatic-runtime = { path = "../peepmatic/crates/runtime", optional = true }
 regalloc = "0.0.24"
+rspirv = "0.6.0"
+spirv_headers = "1.4.2"
```
Comment (bjorn3) [inline cranelift/codegen/Cargo.toml:32]: Please make these optional and depend on them from the `spriv` feature.

File: cranelift/codegen/Cargo.toml [hunk 2]
```diff
@@ -61,14 +64,16 @@ arm32 = []
...
 
 # Option to enable all architectures.
 all-arch = [
     "x86",
     "x64",
     "arm32",
     "arm64",
-    "riscv"
+    "riscv",
+    "spirv"
```
Comment (bjorn3) [inline cranelift/codegen/Cargo.toml:78]: Does toml support trailing commas? If so, please add one to minimize diffs a bit for future architecture additions.

File: cranelift/codegen/meta/src/isa/mod.rs [hunk 3]
```diff
@@ -48,9 +51,11 @@ impl fmt::Display for Isa {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match *self {
             Isa::Riscv => write!(f, "riscv"),
+            Isa::Spirv => write!(f, "spirv"),
```
Comment (bjorn3) [inline cranelift/codegen/meta/src/isa/mod.rs:54]:
```suggestion
```

Already mentioned below.

File: cranelift/codegen/meta/src/isa/mod.rs [hunk 4]
```diff
@@ -7,6 +7,7 @@ mod arm32;
 mod arm64;
 mod riscv;
 mod x86;
+mod spirv;
```
Comment (bjorn3) [inline cranelift/codegen/meta/src/isa/mod.rs:9]: @cfallin Is it still necessary to specify backends in `codegen-meta` when using the new framework?

File: cranelift/codegen/src/isa/mod.rs [hunk 5]
```diff
@@ -83,6 +83,8 @@ mod x64;
 #[cfg(feature = "arm32")]
 mod arm32;
 
+mod spirv;
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/mod.rs:87]:
```suggestion
#[cfg(feature = "spirv")]
mod spirv;
```

File: cranelift/codegen/src/isa/spirv/inst.rs [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    fn emit(&self, sink: &mut MachBuffer<Self>, flags: &Flags, state: &mut Self::State) {
+        let mut operands = vec![];
+        
+        for operand in &self.operands {
+            let idx = operand.get_index();
+            operands.push(idx as u32); 
+        }
+        
+        let mut opcode_len = 1 + operands.len() as u32;
+        if let Some(r) = self.result_type {
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/spirv/inst.rs:84]:
```suggestion
        if self.result_type.is_some() {
```

File: cranelift/codegen/src/isa/spirv/inst.rs [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+            opcode_len += 1;
+        }
+
+        if let Some(r) = self.result_id {
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/spirv/inst.rs:88]:
```suggestion
        if self.result_id.is_some() {
```

File: cranelift/codegen/src/isa/spirv/inst.rs [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        for op in operands {
+            sink.put4(op);
+        }
+    }
+}
+
+
+impl MachInst for Inst {
+    fn get_regs(&self, collector: &mut RegUsageCollector) {
+        if let Some(result) = self.result_id{
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/spirv/inst.rs:112]: nit:

```suggestion
        if let Some(result) = self.result_id {
```

File: cranelift/codegen/src/isa/spirv/inst.rs [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+    ) -> (CodeOffset, LabelUse) {
+        (0, Self::Standard)
+    }
+}
+
+impl ShowWithRRU for Inst {
+    fn show_rru(&self, mb_rru: Option<&RealRegUniverse>) -> String {
+        format!("nop")
+    }
+}
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/spirv/inst.rs:242]: nit: missing trailing newline

**Comment by Jasper-Bekkers** [comment]
Rebased changes on top of mainline now.

File: cranelift/codegen/src/isa/spirv/lower.rs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    fn lower_branch_group<C: LowerCtx<I = Self::MInst>>(
+        &self,
+        ctx: &mut C,
+        insts: &[IRInst],
+        targets: &[MachLabel],
+        fallthrough: Option<MachLabel>,
+    ) -> CodegenResult<()> {
+        Err(CodegenError::Unsupported(format!("lower_branch_group")))
+    }
+}
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/spirv/lower.rs:79]: nit: missing trailing newline

File: cranelift/codegen/src/isa/spirv/mod.rs [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+            triple,
+            flags,
+            reg_universe,
+        }
+    }
+
+    fn compile_vcode(&self, func: &Function, flags: Flags) -> CodegenResult<VCode<Inst>> {
+        // This performs lowering to VCode, register-allocates the code, computes
+        // block layout and finalizes branches. The result is ready for binary emission.
+        debug!("Hoi");
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/spirv/mod.rs:63]: :wave:

File: cranelift/codegen/src/isa/spirv/mod.rs [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    assert!(triple.architecture == Architecture::Spirv);
+    IsaBuilder {
+        triple,
+        setup: settings::builder(),
+        constructor: |triple, shared_flags, _| {
+            let backend = SpirvBackend::new_with_flags(triple, shared_flags);
+            Box::new(TargetIsaAdapter::new(backend))
+        },
+    }
+}
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/spirv/mod.rs:124]: nit: missing trailing newline

File: cranelift/codegen/src/machinst/compile.rs [hunk 13]
```diff
@@ -72,20 +72,19 @@ where
...
                 vcode.show_rru(Some(b.reg_universe())),
                 err
-            );
-            err
-        })
-        .expect("register allocation")
-    };
+            })
+            .expect("register allocation")
+        };
```
Comment (bjorn3) [inline cranelift/codegen/src/machinst/compile.rs:77]: Something went wrong here. Can you revert this change?

File: cranelift/test.clif [hunk 14]
```diff
@@ -0,0 +1,7 @@
+function %add() -> i32 {
+block0:
+    v0 = iconst.i32 40
+    v1 = iconst.i32 2
+    v2 = iadd v0, v1
+    return v2
+}
```
Comment (bjorn3) [inline cranelift/test.clif:7]: nit: missing trailing newline

**Comment by bjorn3** [comment]
I think there should be a way to disable regalloc for the SPIR-V backend.

File: cranelift/Cargo.toml [hunk 15]
```diff
@@ -47,4 +47,4 @@ walkdir = "2.2"
 default = ["disas", "wasm", "cranelift-codegen/all-arch"]
 disas = ["capstone"]
 enable-peepmatic = ["cranelift-codegen/enable-peepmatic", "cranelift-filetests/enable-peepmatic"]
-wasm = ["wat", "cranelift-wasm"]
+wasm = ["wat", "cranelift-wasm"]
```
Comment (bjorn3) [inline cranelift/Cargo.toml:50]: nit: missing trailing newline

File: cranelift/codegen/src/machinst/compile.rs (hunk 13 above)
Comment (Jasper-Bekkers) [reply cranelift/codegen/src/machinst/compile.rs:77]: It looks like some more went wrong - not just here, during the rebase on top of main, investigating.

**Comment by bjorn3** [comment]
The last commit changes the `tests/spec_testsuite` submodule.

File: cranelift/codegen/meta/src/isa/mod.rs (hunk 4 above)
Comment (bnjbvr) [reply cranelift/codegen/meta/src/isa/mod.rs:10]: It should be only useful to require a new-backend style meta build, that is, build the instructions and `expand` group. So the module is not strictly required here, the `Isa` variant below is sufficient, and should be handled in the second loop in `codegen/meta/src/lib.rs`, probably with an empty match arm (as aarch64) if there are no spirv specific settings. Also the `codegen/build.rs` script should mimic what's done for x64.

File: cranelift/codegen/src/machinst/compile.rs (hunk 13 above)
Comment (Jasper-Bekkers) [reply cranelift/codegen/src/machinst/compile.rs]: Re-rebased on `main` this time (didn't catch the branch rename the first time around, but it's all fixed now).

File: cranelift/codegen/Cargo.toml (hunk 2 above)
Comment (Jasper-Bekkers) [reply cranelift/codegen/Cargo.toml]: Seems so, fixed.

File: cranelift/codegen/Cargo.toml (hunk 1 above)
Comment (Jasper-Bekkers) [reply cranelift/codegen/Cargo.toml]: Done.

File: cranelift/codegen/src/isa/mod.rs (hunk 5 above)
Comment (Jasper-Bekkers) [reply cranelift/codegen/src/isa/mod.rs:88]: Changed.

File: cranelift/codegen/src/isa/spirv/inst.rs (hunk 6 above)
Comment (Jasper-Bekkers) [reply cranelift/codegen/src/isa/spirv/inst.rs]: Done.

File: cranelift/codegen/meta/src/isa/spirv/mod.rs [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+    let builder = RegBankBuilder::new("VirtualRegs", "")
+        .units(255) // jb-todo: spirv's registers are virtual...
+        .track_pressure(true);
+    let int_regs = regs.add_bank(builder);
+
+    let builder = RegClassBuilder::new_toplevel("GPR", int_regs);
+    regs.add_class(builder);
+
+    regs.build()
```
Comment (bjorn3) [inline cranelift/codegen/meta/src/isa/spirv/mod.rs:30]: I think this is can be return an empty `IsaRegs`.

File: cranelift/codegen/meta/src/lib.rs [hunk 17]
```diff
@@ -116,7 +116,7 @@ pub fn generate(
             isa::Isa::Arm64 => {
                 // aarch64 doesn't have platform-specific settings.
             }
-            isa::Isa::Arm32 | isa::Isa::Riscv => todo!(),
+            isa::Isa::Arm32 | isa::Isa::Riscv | isa::Isa::Spirv => todo!(),
```
Comment (bjorn3) [inline cranelift/codegen/meta/src/lib.rs:119]: I think `Spirv` can be handled the same as `Arm64`.

**Comment by Jasper-Bekkers** [comment]
> The last commit changes the `tests/spec_testsuite` submodule.

I'll take a look at that - it shouldn't.

File: cranelift/codegen/src/isa/spirv/abi.rs [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+use std::sync::atomic::{AtomicU32, Ordering};
+
+struct SpirvReturn {
+    id: u32,
+    real_reg: RealReg,
+}
+
+pub(crate) struct SpirvABIBody {
+    flags: settings::Flags,
+    next_id: AtomicU32,
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/spirv/abi.rs:25]: This won't work they way you may think it does. You are only accessing it from a single thread, so it is equivalent to an `u32`. What about adding a relocation type for ids, so that the code creating a SPIR-V module from all functions can make the ids unique?

File: cranelift/codegen/src/isa/spirv/inst.rs [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+        }
+
+        for op in operands {
+            sink.put4(op);
+        }
+    }
+
+    fn pretty_print(&self, mb_rru: Option<&RealRegUniverse>, state: &mut EmitState) -> String {
+        use crate::alloc::string::ToString;
+        "".to_string()
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/spirv/inst.rs:117]: Reminder to implement.

File: cranelift/codegen/src/isa/spirv/abi.rs (hunk 18 above)
Comment (Jasper-Bekkers) [reply cranelift/codegen/src/isa/spirv/abi.rs:25]: Yeah; we're having a discussion internally on how to best solve this. We'll need to hand out `id`'s in a few cases; one is to allocate types and the other as virtual SSA registers, so we need some architecture around this. 

For type allocation we're thinking of doing a two-pass approach (eg. scan which types are needed and alloc them followed by a pass that just uses them).

For handing out id's during instruction lowering, we'll need some kind of way to have relocateable id's most likely so we'll take a look at that.

File: cranelift/codegen/src/isa/spirv/inst.rs (hunk 19 above)
Comment (Jasper-Bekkers) [reply cranelift/codegen/src/isa/spirv/inst.rs]: Replaced with `todo!`

File: cranelift/codegen/meta/src/lib.rs (hunk 17 above)
Comment (Jasper-Bekkers) [reply cranelift/codegen/meta/src/lib.rs]: Done.

File: cranelift/codegen/meta/src/isa/spirv/mod.rs (hunk 16 above)
Comment (Jasper-Bekkers) [reply cranelift/codegen/meta/src/isa/spirv/mod.rs]: Done.

**Comment by Jasper-Bekkers** [comment]
Closing this issue because we chose to peruse the same goal through different means.
